src/metabase/api/common.clj

```diff
@@ -329,6 +335,113 @@
+  "Interval between sending newline characters to keep Heroku from terminating requests like queries that take a long
+  time to complete."
+  (* 1 1000))
+
+;; Handle ring response maps that contain a core.async chan in the :body key:
+;;
+;; {:status 200
+;;  :body (async/chan)}
+;;
+;; and send strings (presumibly \n) as heartbeats to the client until the real results (a seq) is received, then
+;; stream that to the client
+(extend-protocol protocols/StreamableResponseBody
+  clojure.core.async.impl.channels.ManyToManyChannel
+  (write-body-to-stream [output-queue _ ^OutputStream output-stream]
+    (log/debug (u/format-color 'green "starting streaming request"))
+    (with-open [out (io/writer output-stream)]
+      (loop [chunk (async/<!! output-queue)]
+        (cond
+          (char? chunk)
+          (do
+            (try
+              (.write out (str chunk))
+              (.flush out)
+              (catch org.eclipse.jetty.io.EofException e
+                (log/info e (u/format-color 'yellow "connection closed, canceling request"))
+                (async/close! output-queue)
+                (throw e)))
+            (recur (async/<!! output-queue)))
+
+          ;; An error has occurred, let the user know
+          (instance? Exception chunk)
+          (json/generate-stream {:error (.getMessage ^Exception chunk)} out)
+
+          ;; We've recevied the response, write it to the output stream and we're done
+          (seq chunk)
+          (json/generate-stream chunk out)
+
+          ;;chunk is nil meaning the output channel has been closed
+          :else
+          out)))))
+
+(def ^:private InvokeWithKeepAliveSchema
+  {;; Channel that contains any number of newlines followed by the results of the invoked query thunk
+   :output-channel  (s/protocol async-proto/Channel)
+   ;; This channel will have an exception if that error condition is hit before the first heartbeat time, if a
+   ;; heartbeat has been sent, this channel is closed and its no longer useful
+   :error-channel   (s/protocol async-proto/Channel)
+   ;; Future that is invoking the query thunk. This is mainly useful for testing metadata to see if the future has been
+   ;; cancelled or was completed successfully
+   :response-future java.util.concurrent.Future})
+
+(s/defn ^:private invoke-thunk-with-keepalive :- InvokeWithKeepAliveSchema
+  "This function does the heavy lifting of invoking `query-thunk` on a background thread and returning it's results
+  along with a heartbeat while waiting for the results. This function returns a map that includes the relevate
+  execution information, see `InvokeWithKeepAliveSchema` for more information"
+  [query-thunk]
+  (let [response-chan (async/chan 1)
+        output-chan   (async/chan 1)
+        error-chan    (async/chan 1)
+        response-fut  (future
+                        (try
+                          (async/>!! response-chan (query-thunk))
+                          (async/close! error-chan)
```

Comment:
Lift to finally since it seems to be an invariant?

Reply:
:+1: